client/src/telethon/_impl/mtsender/sender.py

```diff
@@ -162,18 +163,20 @@ class Request(Generic[Return]):
 class Sender:
     dc_id: int
     addr: str
-    lock: Lock
     _logger: logging.Logger
+    _lock: Lock
     _reader: AsyncReader
     _writer: AsyncWriter
     _transport: Transport
     _mtp: Mtp
     _mtp_buffer: bytearray
-    _requests: list[Request[object]]
-    _request_event: Event
+    _send_queue: Queue[Request[object]]
+    _recv_queue: Queue[Request[object]]
+    _updates_queue: Queue[Updates]
     _next_ping: float
     _read_buffer: bytearray
-    _write_drain_pending: bool
+    _send_task: Task | None = None
+    _recv_task: Task | None = None
```

Comment:
I appreciate your time, however, and sorry for the back and forth, the avoidance of tasks was explicit. The excellent [Notes on structured concurrency](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/) explain it better than I could, but in short, not having background tasks greatly simplifies cancellation and makes the flow explicit.

I would strongly prefer if we could avoid them as much as possible.

We *might* want to spawn background tasks in the client, to handle updates in the background and sending periodic pings, but I would strongly prefer to keep the internals as clean as possible.

The design of this sender is inspired by grammers, which also avoids creating background tasks *entirely*, and I think it's a nice property. Certainly, it requires a lot more care and work, but I believe it's worthwhile.

There must be another way to solve the issue you mentioned in https://github.com/LonamiWebs/Telethon/pull/4484#issuecomment-2408642344.

Perhaps `self._reader.read(MAXIMUM_DATA)` is not cancel-safe and dropping it causes data to be lost. But that wouldn't explain the error either.